lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -94,105 +94,180 @@ void ApplyRewriteMapOnResTable(RemapEntryCollection &rewrites, DxilModule &DM) {
+    unsigned AutoBindingSpace) {
+    bool bChanged = false;
+    SpacesAllocator<unsigned, T> SAlloc;
+
+    // Reserve explicitly allocated resources
+    for (auto &res : resourceList) {
+      const unsigned space = res->GetSpaceID();
+      typename SpacesAllocator<unsigned, T>::Allocator &alloc = SAlloc.Get(space);
+      typename SpacesAllocator<unsigned, T>::Allocator &reservedAlloc = ReservedRegisters.Get(space);
+
+      if (res->IsAllocated()) {
+        const unsigned reg = res->GetLowerBound();
+        const T *conflict = nullptr;
+        if (res->IsUnbounded()) {
+          const T *unbounded = alloc.GetUnbounded();
+          if (unbounded) {
+            Ctx.emitError(Twine("more than one unbounded resource (") +
+              unbounded->GetGlobalName() + (" and ") +
+              res->GetGlobalName() + (") in space ") + Twine(space));
+          }
+          else {
+            conflict = alloc.Insert(res.get(), reg, res->GetUpperBound());
+            if (!conflict) {
+              alloc.SetUnbounded(res.get());
+              reservedAlloc.SetUnbounded(res.get());
+            }
+          }
+        }
+        else {
           conflict = alloc.Insert(res.get(), reg, res->GetUpperBound());
-          if (!conflict)
-            alloc.SetUnbounded(res.get());
         }
-      } else {
-        conflict = alloc.Insert(res.get(), reg, res->GetUpperBound());
+        if (conflict) {
+          Ctx.emitError(((res->IsUnbounded()) ? Twine("unbounded ") : Twine("")) +
+            Twine("resource ") + res->GetGlobalName() +
+            Twine(" at register ") + Twine(reg) +
+            Twine(" overlaps with resource ") +
+            conflict->GetGlobalName() + Twine(" at register ") +
+            Twine(conflict->GetLowerBound()) + Twine(", space ") +
+            Twine(space));
+        }
+        else {
+          // Also add this to the reserved (unallocatable) range, if it wasn't already there.
+          reservedAlloc.ForceInsertAndClobber(res.get(), res->GetLowerBound(), res->GetUpperBound());
+        }
       }
-      if (conflict) {
-        Ctx.emitError(((res->IsUnbounded()) ? Twine("unbounded ") : Twine("")) +
-                      Twine("resource ") + res->GetGlobalName() +
-                      Twine(" at register ") + Twine(reg) +
-                      Twine(" overlaps with resource ") +
-                      conflict->GetGlobalName() + Twine(" at register ") +
-                      Twine(conflict->GetLowerBound()) + Twine(", space ") +
-                      Twine(space));
+    }
+
+    // Allocate unallocated and bounded resources
+    const unsigned space = AutoBindingSpace;
+    typename SpacesAllocator<unsigned, T>::Allocator &alloc0 = SAlloc.Get(space);
+    typename SpacesAllocator<unsigned, T>::Allocator &reservedAlloc0 = ReservedRegisters.Get(space);
+    for (auto &res : resourceList) {
+      if (!res->IsAllocated() && !res->IsUnbounded()) {
```

Comment:
There might be more changes here than is needed. I split the loop for unallocated resources between bounded first and then unbounded resources because I thought I was hitting an issue where the below would start failing, but I'm not sure anymore. I don't see how the original code could have handled this case.

```
Texture2D foo[]; // expected: t1 to t-infinity?
Texture2D bar : register(t0);
```

Reply:
Order of collection/reservation should be based on explicit register binding, then order of declaration, not whether or not it's an unbounded array.  And we should disallow use of unbounded resources with -flegacy-resource-reservations, since they were introduced in SM 5.1.

This case would pass originally because: if bar was used, it would get reserved first, since it's explicitly bound, then foo would get allocated after bar.  If bar was unused, it would get removed earlier, and foo would get allocated on it's own, starting from t0.